Improve error reporting for case files.

AiredaleDev opened the pull request:
## Description

If a case file's parameter is assigned a value of the wrong type, MFC will now correctly report the mismatching field. Previously, it would point to the next parameter, even if it typechecks.

Fixes #424

### Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

### Scope

- [x] This PR comprises a set of related changes with a common goal

If you cannot check the above box, please split your PR into multiple PRs that each have a common goal.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. 
Provide instructions so we can reproduce. 
Please also list any relevant details for your test configuration

- [x] Assertions that ensure that typechecking logic covers all parameters
- [x] Modifying and running a case file with (in)correct values.

**Test Configuration**:

* What computers and compilers did you use to test this:
My personal laptop:
> Ryzen 7 7840U
> GCC 11.4.0

## Checklist

- [x] I have added comments for the new code
- [x] I have made corresponding changes to the documentation (`docs/`) -- (I caught a typo while grepping through the codebase: Loginal -> Logical)
- [x] I have added regression tests to the test suite so that people can verify in the future that the feature is behaving as expected
- [x] I ran `./mfc.sh format` before committing my code
- [x] New and existing tests pass locally with my changes, including with GPU capability enabled (both NVIDIA hardware with NVHPC compilers and AMD hardware with CRAY compilers) and disabled
- [x] This PR does not introduce any repeated code (it follows the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle)
- [x] I cannot think of a way to condense this code and reduce any introduced additional line count -- I considered an automated solution to get the types from the Fortran codebase that uses `grep` and `awk`, but it seemed too brittle and complicated.

sbryngelson commented:
This _seems_ right to me, but the rightful person to review this is @henryleberre... nice job @AiredaleDev

AiredaleDev commented:
Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?

AiredaleDev commented:
> This _seems_ right to me, but the rightful person to review this is @henryleberre... nice job @AiredaleDev

Thanks! Henry and I reviewed it last night, there's only the one outstanding thing I just discovered before it should be ready to merge.

sbryngelson commented:
> Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?

Can you be a bit more precise... how long are we talking? 

and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?

AiredaleDev commented:
> > Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?
> 
> Can you be a bit more precise... how long are we talking?
> 
> and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?

Roughly a minute or two, though it's only for `./mfc.sh test` since it generates a lot of `MFCInputFile`s `./mfc.sh build` and `./mfc.sh run` are fine because they validate at most one case file.

sbryngelson commented:
> > > Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?
> > 
> > 
> > Can you be a bit more precise... how long are we talking?
> > and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?
> 
> Roughly a minute or two, though it's only for `./mfc.sh test` since it generates a lot of `Case`s. `./mfc.sh build` and `./mfc.sh run` are fine because they validate at most one case file.

That is a lot. Can we switch to type checking at every test initialization rather than all at once in the beginning? That would probably fix it... distributing 1 minute over 150 tests makes it negligible. Need to be careful with the multithreading going on, through `./mfc.sh test -j X`

AiredaleDev commented:
> > > > Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?
> > > 
> > > 
> > > Can you be a bit more precise... how long are we talking?
> > > and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?
> > 
> > 
> > Roughly a minute or two, though it's only for `./mfc.sh test` since it generates a lot of `Case`s. `./mfc.sh build` and `./mfc.sh run` are fine because they validate at most one case file.
> 
> That is a lot. Can we switch to type checking at every test initialization rather than all at once in the beginning? That would probably fix it... distributing 1 minute over 150 tests makes it negligible. Need to be careful with the multithreading going on, through `./mfc.sh test -j X`

We appear to be already doing that *in addition* to the initial check. Removing these means we'll have to go back to waiting until after MFC is built to validate these inputs, but I think that's fine.

sbryngelson commented:
> > > > > Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?
> > > > 
> > > > 
> > > > Can you be a bit more precise... how long are we talking?
> > > > and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?
> > > 
> > > 
> > > Roughly a minute or two, though it's only for `./mfc.sh test` since it generates a lot of `Case`s. `./mfc.sh build` and `./mfc.sh run` are fine because they validate at most one case file.
> > 
> > 
> > That is a lot. Can we switch to type checking at every test initialization rather than all at once in the beginning? That would probably fix it... distributing 1 minute over 150 tests makes it negligible. Need to be careful with the multithreading going on, through `./mfc.sh test -j X`
> 
> We appear to be already doing that _in addition_ to the initial check. Removing the initial check means we'll have to go back to waiting until after MFC is built to validate the case inputs, but I think that's fine since the alternative is sequentially validating all 150 inputs. Each test has a unique file that it reads iirc, so multithreading should behave fine.
> 
> I also just discovered we could validate only the changed parameters upfront, but the test runner is already going to run the check used during a normal run and the one it uses for generating its inputs, so it would ultimately do redundant work.

I don't know what you mean by it already does that, can you be more specific?

MFC is only built once before the test suite runs.

I think trying to get the toolchain to only check for changed inputs is more complex than appropriate for this.

AiredaleDev commented:
> > > > > > Adding the check in the constructor significantly extends the pre-build time for the test suite, enough for the user to think something is wrong (validating json is apparently quite expensive). For performance reasons, I may have to remove the check in the constructor, is there a way we can run fewer checks?
> > > > > 
> > > > > 
> > > > > Can you be a bit more precise... how long are we talking?
> > > > > and I guess you mean it expands the time between initiating `./mfc.sh build` and it actually starting to build? or `./mfc.sh run` and it actually starting to run?
> > > > 
> > > > 
> > > > Roughly a minute or two, though it's only for `./mfc.sh test` since it generates a lot of `Case`s. `./mfc.sh build` and `./mfc.sh run` are fine because they validate at most one case file.
> > > 
> > > 
> > > That is a lot. Can we switch to type checking at every test initialization rather than all at once in the beginning? That would probably fix it... distributing 1 minute over 150 tests makes it negligible. Need to be careful with the multithreading going on, through `./mfc.sh test -j X`
> > 
> > 
> > We appear to be already doing that _in addition_ to the initial check. Removing the initial check means we'll have to go back to waiting until after MFC is built to validate the case inputs, but I think that's fine since the alternative is sequentially validating all 150 inputs. Each test has a unique file that it reads iirc, so multithreading should behave fine.
> > I also just discovered we could validate only the changed parameters upfront, but the test runner is already going to run the check used during a normal run and the one it uses for generating its inputs, so it would ultimately do redundant work.
> 
> I don't know what you mean by it already does that, can you be more specific?
> 
> MFC is only built once before the test suite runs.
> 
> I think trying to get the toolchain to only check for changed inputs is more complex than appropriate for this.

Agreed. If I were to go about it, I'd have to move check to two unintuitive places to perform a full check in the case of `run` and only a partial, upfront check in the case of `test` (call `validate` on the `mods` parameter of `TestCase`'s constructor). Moving the check into `MFCInputFile`'s constructor saves us the upfront minute, so I'll call it done.

sbryngelson commented:
If I understand correctly, you have it setup so it builds _all_ of the test cases and then checks _all_ of the parameters for _all_ the tests before actually starting running the tests, whereas the existing (pre-PR) implementation didn't check the parameters of the test cases at all? or did the pre-PR implementation only test some of them somehow? I'm not more confused about the nature of the issue #424

AiredaleDev commented:
> If I understand correctly, you have it setup so it builds _all_ of the test cases and then checks _all_ of the parameters for _all_ the tests before actually starting running the tests, whereas the existing (pre-PR) implementation didn't check the parameters of the test cases at all? or did the pre-PR implementation only test some of them somehow? I'm now more confused about the nature of the issue #424

Okay, I'm not explaining it right. The parameters were initially validated in the Fortran codebase, but the solution was quite hacky and would often point to benign parameters. In this PR, each run now checks all of its parameters using the schema, which provides clear error reporting, and then launches MFC if everything is okay -- this is true regarding normal runs and test suite.

I unintentionally made a change that would do as you described, which introduced the minute-long delay in the test suite.  I fixed this delay with my most recent commits.

sbryngelson commented:
ah - got it! I understand now.

henryleberre commented:
@AiredaleDev @sbryngelson I think the best approach here is to define a method under `Case` called `validate()`. Code that makes use of `Case` objects should have the burden of calling it as a sanity check.

henryleberre commented:
Also @AiredaleDev, I am to blame for the inclusion of (most of) the unused variables you found (`chem_wrt`, `spec_pp`, ...). These were unintentionally added when upstreaming some changes from my branch that is adding chemistry.

AiredaleDev commented:
> @AiredaleDev @sbryngelson I think the best approach here is to define a method under `Case` called `validate()`. Code that makes use of `Case` objects should have the burden of calling it as a sanity check.

@henryleberre Normally I would agree, but if I apply this sanity check to all code that constructs a `Case` (i.e. also a `TestCase`) then this line in the test runner forces the user to wait almost one whole minute before it begins. Apparently `jsonschema` is a bit slow.

```py
# In toolchain/mfc/test/test.py
def test():
    # pylint: disable=global-statement, global-variable-not-assigned
    global nFAIL

    # This is the line I have to work around.
    # It runs before anything in parallel happens, so all 150 cases are checked sequentially.
    cases = [ _.to_case() for _ in list_cases() ]

# In toolchain/mfc/test/case.py
    def to_case(self) -> TestCase:
        dictionary = self.mods.copy()
        if self.path:
            dictionary.update(input.load(self.path, self.args).params)

            for key, value in dictionary.items():
                if not isinstance(value, str):
                    continue

                for path in [value, os.path.join(os.path.dirname(self.path), value)]:
                    path = os.path.abspath(path)
                    if os.path.exists(path):
                        dictionary[key] = path
                        break
        # This code "should" check the dictionary, but if we check the dictionary here
        # we wait ~45 seconds before the test runner starts.
        return TestCase(self.trace, dictionary, self.ppn, self.rebuild)
```

The test runner will eventually call `run` anyway, which ends up reading and validating each case dictionary in parallel before launching the simulation (i.e. each concurrent run of MFC checks its own parameters.)

I added a `validate` method with the last commit, but it's only used in the same place its body was used before. Regarding @sbryngelson 's request, I added a case to `examples` that intentionally triggers the error reporting when run.

sbryngelson commented:
hi @AiredaleDev no need to actually include broken code in version control, just 'proof' or 'verification' that if you do something like this your code identifies it.

sbryngelson commented:
not sure I quite follow @AiredaleDev -- I know `test` is just wrapper around `run` at some level, so I think @henryleberre is saying that everything that creates a case, which would be any `run` call, including those invoked through `test`, gets its case file validated. I don't think `test` should have any special machinery for this but rather invoke it at the `run` level?

henryleberre commented:
> > Code that makes use of `Case` objects should have the burden of calling it as a sanity check.
> 
> @henryleberre Normally I would agree, but if I apply this sanity check to all code that constructs a `Case` (i.e. also a `TestCase`) then this line in the test runner forces the user to wait almost one whole minute before it begins. Apparently `jsonschema` is a bit slow.
> 
> The test runner will eventually call `run` anyway, which ends up reading and validating each case dictionary in parallel before launching the simulation (i.e. each concurrent run of MFC checks its own parameters then runs if they are correct.) I discovered this by adding a print statement next to the validation code, and if you'd like to check this yourself, add a print statement to the newly added `validate` method.
> 
> ```python
> # In toolchain/mfc/test/test.py
> def test():
>     # pylint: disable=global-statement, global-variable-not-assigned
>     global nFAIL
> 
>     # This is the line I have to work around.
>     # It runs before anything in parallel happens, so all 150 cases are checked sequentially.
>     cases = [ _.to_case() for _ in list_cases() ]
> 
> # In toolchain/mfc/test/case.py
>     def to_case(self) -> TestCase:
>         dictionary = self.mods.copy()
>         if self.path:
>             dictionary.update(input.load(self.path, self.args).params)
> 
>             for key, value in dictionary.items():
>                 if not isinstance(value, str):
>                     continue
> 
>                 for path in [value, os.path.join(os.path.dirname(self.path), value)]:
>                     path = os.path.abspath(path)
>                     if os.path.exists(path):
>                         dictionary[key] = path
>                         break
>         # This code "should" check the dictionary, but if we check the dictionary here
>         # we wait ~45 seconds before the test runner starts.
>         return TestCase(self.trace, dictionary, self.ppn, self.rebuild)
> 
> # In toolchain/mfc/test/test.py
> def test():
>     # ...
> 
>     sched.sched(sched.Task(..., func=handle_func, ...)
> 
> # Called by handle_func
> def _handle_func():
>    # ...
>    case.run(...) # case : TestCase
> 
> class TestCase(case.Case):
>     # ...
>     def run(...):
>         # ...
> 
>         # We just end up calling `./mfc.sh run` for each test anyway.
>         command = [
>             mfc_script, "run", filepath, *rebuild, *tasks, *case_optimization,
>             *jobs, "-t", *target_names, *gpus_select, *ARG("--")
>         ]
> 
>         return common.system(command, print_cmd=False, text=True, capture_output=True)
> ```
> 
> Regarding @sbryngelson 's request, I added a case to `examples` that intentionally triggers the error reporting.

You're right. Sorry if I was unclear. I believe we agree on this. When I said that:
> Code that makes use of `Case` objects should have the burden of calling it as a sanity check.

I did not intend to imply that code that makes use of Case objects should always call `validate`. This decoupling of validation & construction is intended to facilitate this. You only run `validate` when it is deemed reasonable / necessary. I agree we shouldn't call it in `./mfc.sh test` and just let the spawned `./mfc.sh run` processes do the validation.

AiredaleDev commented:
> not sure I quite follow @AiredaleDev -- I know `test` is just wrapper around `run` at some level, so I think @henryleberre is saying that everything that creates a case, which would be any `run` call, including those invoked through `test`, gets its case file validated. I don't think `test` should have any special machinery for this but rather invoke it at the `run` level?

This is how it behaves currently.

> I did not intend to imply that code that makes use of Case objects should always call validate. This decoupling of validation & construction is intended to facilitate this. You only run validate when it is deemed reasonable / necessary. I agree we shouldn't call it in ./mfc.sh test and just let the spawned ./mfc.sh run processes do the validation.

Ah, I see! I'll go in and address that real quick.

sbryngelson commented:
@henryleberre can approve and merge this PR when it's ready

AiredaleDev commented:
> hi @AiredaleDev no need to actually include broken code in version control, just 'proof' or 'verification' that if you do something like this your code identifies it.

Okay, will remove the broken case and show you the logs:

henryleberre commented on toolchain/mfc/run/case_dicts.py:
```diff
@@ -1,5 +1,45 @@
 from ..state import ARG
 
+# Currently doing this manually -- the first automated solution I came up with involved grep and awk, but it seemed quite brittle.
+INTEGERS = {
+    "m", "n", "p", 'thermal', 't_step_old', 't_step_start', 't_step_stop',
+    't_step_save', 'num_patches', 'model_eqns', 'num_fluids', "weno_order",
+    "precision", "relax_model", 'perturb_flow_fluid', 'perturb_sph_fluid',
+    'dist_type', 'R0_type', 'num_ibs', 't_step_print', 'time_stepper',
+    'riemann_solver', 'wave_speeds', 'avg_state', 'fd_order', 'num_probes',
+    'bubble_model', 'num_mono', 'num_integrals', 'format', 'flux_lim',
+}
+REALS = {
+    "pref", "Web", "poly_sigma", 'x_domain%beg', 'x_domain%end', 'y_domain%beg',
+    'y_domain%end', 'dt', "nb", "rhoref", "R0ref", "Re_inv", "Ca",
+    'perturb_flow_mag', 'sigR', 'sigV', 'rhoRV', "palpha_eps", "ptgalpha_eps",
+    'pi_fac', 'weno_eps', 'schlieren_alpha', 'fluid_rho',
+}
+LOGICALS = {
+    "hypoelasticity", "cyl_coord", "parallel_io", "run_time_info", "bubbles",
+    "polytropic", "polydisperse", "file_per_process", "relax", "mpp_lim", 
+    "adv_alphan",  "adv_n", 'old_grid', 'old_ic', 'instability_wave',
+    'perturb_flow', 'perturb_sph', 'qbmm', 'ib', 'mapped_weno', 'mp_weno',
+    'weno_avg', 'weno_Re_flux', 'prim_vars_wrt', 'alt_soundspeed', 
+    'null_weights', 'probe_wrt', 'Monopole', 'integral_wrt', 'cu_mpi', 
+    'adap_dt', 'schlieren_wrt', 'alpha_rho_wrt', 'rho_wrt', 'mom_wrt', 
+    'vel_wrt', 'flux_wrt', 'E_wrt', 'pres_wrt', 'alpha_wrt', 'kappa_wrt',
+    'gamma_wrt', 'heat_ratio_wrt', 'pi_inf_wrt', 'pres_inf_wrt',
+    'cons_vars_wrt', 'c_wrt', 'omega_wrt', 'qm_wrt', 'vel_profile',
+    'mixture_err',
+}
+STRINGS = {"case_dir"}
```
@AiredaleDev I think you can just initialize these as empty now.

AiredaleDev replied:
This actually prompted me to refactor one more time since I found a way I could easily regex-replace what I currently have into something simpler.

henryleberre commented on toolchain/mfc/run/case_dicts.py:
```diff
@@ -1,5 +1,45 @@
+STRINGS = {"case_dir"}
+
+
+def add_param(param_name: str, stage_list: list[str], type_set: set[str]):
+    '''Arguments:
+    stage_list -- List containing parameters used for a particular stage. 
+                  One of COMMON, PRE_PROCESS, SIMULATION, POST_PROCESS.
+    type_set   -- Set containing parameters of a certain type. Currently
+                  one of INTEGERS, REALS, LOGICALS, STRINGS.'''
+    stage_list.append(param_name)
+    type_set.add(param_name)
```
Perhaps now is a good time to remove the `COMMON` dictionary. You can modify this function such that you can specify multiple `stage_list`s.

AiredaleDev replied:
I ultimately came up with another, better way of representing the types -- you now only have to type them once, resulting in less overall code -- this function is now replaced with a dictionary assignment

henryleberre commented on toolchain/mfc/run/case_dicts.py:
```diff
@@ -20,175 +60,221 @@
+    elif param in STRINGS:
+        _properties[param] = {"type": "string"}
+    else:
+        print(f'WARNING: Found parameter {param} without type!')
```
I think we should throw an error here instead of just a warning.

AiredaleDev replied:
100% agree

henryleberre commented on toolchain/mfc/test/test.py:
```diff
@@ -99,6 +99,7 @@ def test():
 
     for case in cases:
         if case.rebuild:
+            case.validate_params()
```
I think you should move this to `build()`.

henryleberre commented on toolchain/mfc/test/test.py:
```diff
@@ -192,6 +193,7 @@ def _handle_case(case: TestCase, devices: typing.Set[int]):
             if not os.path.exists(silo_filepath):
                 silo_filepath = os.path.join(case.get_dirpath(), 'silo_hdf5', 'p_all', 'p0', f'{t_step}.silo')
 
+            case.validate_params()
```
Any particular reason for checking this here?

AiredaleDev replied:
None in particular -- will see if I can get away without it.

AiredaleDev commented:
@henryleberre Just a reminder that this should be ready to merge!

henryleberre reviewed (approved):
Thanks for making these changes! lgtm.
